### PR TITLE
GAPI Fluid: SIMD SSE41 for Resize F32C1.

### PR DESCRIPTION
**Pull request by anna-khakimova**
SIMD SSE41 for the Resize F32C1.

<cut/>

Performance:
[ResizeF32C1_SIMD_SSE.xlsx](https://github.com/opencv/opencv/files/8222188/ResizeF32C1_SIMD_SSE.xlsx)

```
force_builders=Linux AVX2,Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
Xbuild_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

build_image:Custom=centos:7
buildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

Xbuild_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.2.0

buildworker:Custom Win=windows-3

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
Xtest_opencl:Custom=OFF
Xtest_bigdata:Custom=1
Xtest_filter:Custom=*

CPU_BASELINE:Custom Win=AVX512_SKX
CPU_BASELINE:Custom=SSE4_2
```

### CODE REVIEW

File: modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp [hunk 1]
```diff
@@ -554,7 +554,162 @@ CV_ALWAYS_INLINE void calcRowLinear_8UC_Impl_<3>(uint8_t* dst[],
...
 }
-} // namespace sse42
+
+CV_ALWAYS_INLINE void v_gather_pairs(const float src[], const int* mapsx,
+                                     v_float32x4& low, v_float32x4& high)
+{
+    __m128 l = _mm_setzero_ps();
+    l = _mm_loadl_pi(l, (const __m64*)&src[mapsx[0]]);  // pair of floats
+    l = _mm_loadh_pi(l, (const __m64*)&src[mapsx[1]]);
+    low.val = l;
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp:564]: is it possible to load _mm_loadl_pi/_mm_loadh_pi directly into `low.val`  and 'high.val'?

File: modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp [hunk 2]
```diff
@@ -554,7 +554,162 @@ CV_ALWAYS_INLINE void calcRowLinear_8UC_Impl_<3>(uint8_t* dst[],
...
+                v_float32x4 alpha0 = v_load(&alpha[x]);
+                //  v_float32 alpha1 = 1.f - alpha0;
+
+                v_gather_pairs(src0[line], &mapsx[x], low1, high1);
+                v_deinterleave(low1, high1, s00, s01);
+
+                //  v_float32 res0 = s00*alpha0 + s01*alpha1;
+                v_float32x4 res0 = v_fma(s00 - s01, alpha0, s01);
+
+                v_gather_pairs(src1[line], &mapsx[x], low2, high2);
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp:618]: If you do not rely on compiler optimization still  then don't you think to rearrange operations like  this?
```
gather_pairs(...low1, high1);
gather_pairs(...low2, high2)
v_deinterleave
v_deinterleave
...
v_fma
v_fma
```
it *may* exclude  cache pressure on mapsx

File: modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp [hunk 3]
```diff
@@ -554,7 +554,162 @@ CV_ALWAYS_INLINE void calcRowLinear_8UC_Impl_<3>(uint8_t* dst[],
...
+    else
+    {
+        GAPI_DbgAssert(xRatioEq1 && yRatioEq1);
+        int length = inSz.width;  // == outSz.width
+        for (int line = 0; line < lpi; ++line)
+        {
+            memcpy(dst[line], src0[line], length * sizeof(float));
+        }
+    }
+}
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp:710]: Could you remove unusable comments from this function please

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 4]
```diff
@@ -1930,7 +1934,7 @@ static inline void initScratchLinear(const cv::GMatDesc& in,
         alpha[x] = alpha0;
         index[x] = index0;
 
-        for (int l = 0; l < 4; l++) {
+        for (int l = 0; l < 4; ++l) {
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp:1937]: how much gain does it provide?

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 5]
```diff
@@ -1980,11 +1992,38 @@ struct Mapper {
         return u;
     }
 };
+}  // namespace linear
+
+namespace linear32f {
+struct Mapper {
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp:1998]: looks familiar https://github.com/opencv/opencv/pull/21678/files#diff-7f0bda22bce310481b13ffb6c7114b9c957d864de2b95530dbe994b83c577d5bR1996

Is it kind of duplication?

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 6]
```diff
@@ -2088,32 +2207,44 @@ GAPI_FLUID_KERNEL(GFluidResize, cv::gapi::imgproc::GResize, true)
...
+        else if (depth == CV_32F && channels == 1)
         {
-            calcRowLinearC<uint8_t, Mapper, 3>(in, out, scratch);
+            calcRowLinear<linear32f::Mapper>(in, out, scratch);
         }
-        else if (channels == 4)
+        else
         {
-            calcRowLinearC<uint8_t, Mapper, 4>(in, out, scratch);
+            CV_Error(cv::Error::StsBadArg, "unsupported combination of type and number of channel");
```
Comment (sivanov-work) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp:2247]: Is there a way to report what actually combination failed in troubleshooting purpoese?

File: modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp [hunk 7]
```diff
@@ -554,7 +554,162 @@ CV_ALWAYS_INLINE void calcRowLinear_8UC_Impl_<3>(uint8_t* dst[],
...
+                int   sx1 = sx0 + 1;
+                dst[line][x] = src0[line][sx0] * alpha0 + src0[line][sx1] * alpha1;
+            }
+        }
+
+    }
+    else if (!yRatioEq1)
+    {
+        GAPI_DbgAssert(xRatioEq1);
+        int length = inSz.width;  // == outSz.width
```
Comment (alexgiving) [inline modules/gapi/src/backends/fluid/gfluidcore_simd_sse41.hpp:677]: Is it necessary to keep `length` variable? If so, might it be worth using it in previous cases as well?

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp (hunk 5 above)
Comment (anna-khakimova) [reply modules/gapi/src/backends/fluid/gfluidimgproc.cpp:1998]: This patch is the successor of  the `GAPI Fluid: Resize F32C1 scalar version. `#21678. That is, it contains that patch plus the SIMD add-on.

**Comment by asmorkalov** [comment]
@anna-khakimova friendly reminder.

**Comment by TolyaTalamanov** [comment]
> @TolyaTalamanov Is the patch still relevant? If so, could you drive it to merge by yourself?

@dmatveev Could you comment? I don't recall for what we do this

**Comment by asmorkalov** [comment]
@dmatveev Friendly reminder.

**Comment by dmatveev** [comment]
@asmorkalov I'll have a look on this. At least will actualize the patch and see what effect it gives for F32 resize.

**Comment by dmatveev** [comment]
@rgarnov could you please have a look?

**Comment by dmatveev** [comment]
Checked performance of 4.x against this branch:

```
dm@irlvmvdmitryma:~/code/opencv_fluid/modules/ts/misc$ python3 ./summary.py  ~/code/opencv_fluid_BUILD/4x.xml ~/code/opencv_fluid_BUILD/anna_simd.xml | grep 32FC1
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.5, 0.5, { gapi.kernel_package })                  0.012  0.012     0.97
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.5, 0.25, { gapi.kernel_package })                 0.007  0.008     0.95
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.5, 2, { gapi.kernel_package })                    0.037  0.038     0.96
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.25, 0.5, { gapi.kernel_package })                 0.008  0.008     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.25, 0.25, { gapi.kernel_package })                0.006  0.006     0.97
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 0.25, 2, { gapi.kernel_package })                   0.023  0.023     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 2, 0.5, { gapi.kernel_package })                    0.033  0.033     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 2, 0.25, { gapi.kernel_package })                   0.018  0.018     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 128x128, 2, 2, { gapi.kernel_package })                      0.120  0.120     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.5, 0.5, { gapi.kernel_package })                  0.148  0.148     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.5, 0.25, { gapi.kernel_package })                 0.075  0.075     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.5, 2, { gapi.kernel_package })                    0.557  0.550     1.01
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.25, 0.5, { gapi.kernel_package })                 0.088  0.088     0.99
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.25, 0.25, { gapi.kernel_package })                0.043  0.044     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 0.25, 2, { gapi.kernel_package })                   0.298  0.298     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 2, 0.5, { gapi.kernel_package })                    0.534  0.535     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 2, 0.25, { gapi.kernel_package })                   0.271  0.270     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 640x480, 2, 2, { gapi.kernel_package })                      2.119  2.117     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.5, 0.5, { gapi.kernel_package })                 0.416  0.423     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.5, 0.25, { gapi.kernel_package })                0.212  0.216     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.5, 2, { gapi.kernel_package })                   1.639  1.625     1.01
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.25, 0.5, { gapi.kernel_package })                0.233  0.236     0.99
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.25, 0.25, { gapi.kernel_package })               0.118  0.122     0.97
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 0.25, 2, { gapi.kernel_package })                  0.844  0.842     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 2, 0.5, { gapi.kernel_package })                   1.599  1.594     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 2, 0.25, { gapi.kernel_package })                  0.794  0.797     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1280x720, 2, 2, { gapi.kernel_package })                     6.512  6.520     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.5, 0.5, { gapi.kernel_package })                1.008  1.035     0.97
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.5, 0.25, { gapi.kernel_package })               0.477  0.476     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.5, 2, { gapi.kernel_package })                  3.821  4.011     0.95
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.25, 0.5, { gapi.kernel_package })               0.596  0.605     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.25, 0.25, { gapi.kernel_package })              0.267  0.274     0.98
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 0.25, 2, { gapi.kernel_package })                 2.046  2.216     0.92
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 2, 0.5, { gapi.kernel_package })                  3.697  3.834     0.96
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 2, 0.25, { gapi.kernel_package })                 1.819  1.818     1.00
TestPerformance::ResizeFxFyPerfTestFluid/ResizeFxFyPerfTest::(compare_f, 32FC1, 1, 1920x1080, 2, 2, { gapi.kernel_package })                    14.901 14.888    1.00
TestPerformance::ResizeInSimpleGraphPerfTestFluid/ResizeInSimpleGraphPerfTest::(compare_f, 32FC1, 128x128, 0.5, 0.5, { gapi.kernel_package })   0.026  0.024     1.07
TestPerformance::ResizeInSimpleGraphPerfTestFluid/ResizeInSimpleGraphPerfTest::(compare_f, 32FC1, 640x480, 0.5, 0.5, { gapi.kernel_package })   0.278  0.272     1.02
TestPerformance::ResizeInSimpleGraphPerfTestFluid/ResizeInSimpleGraphPerfTest::(compare_f, 32FC1, 1280x720, 0.5, 0.5, { gapi.kernel_package })  0.809  0.792     1.02
TestPerformance::ResizeInSimpleGraphPerfTestFluid/ResizeInSimpleGraphPerfTest::(compare_f, 32FC1, 1920x1080, 0.5, 0.5, { gapi.kernel_package }) 2.246  2.214     1.01
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 128x128, 30x30, { gapi.kernel_package })                             0.005  0.006     0.96
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 128x128, 64x64, { gapi.kernel_package })                             0.011  0.012     0.97
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 640x480, 30x30, { gapi.kernel_package })                             0.007  0.007     0.98
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 640x480, 64x64, { gapi.kernel_package })                             0.014  0.014     0.97
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 1280x720, 30x30, { gapi.kernel_package })                            0.008  0.008     0.99
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 1280x720, 64x64, { gapi.kernel_package })                            0.018  0.017     1.10
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 1920x1080, 30x30, { gapi.kernel_package })                           0.009  0.009     0.99
TestPerformance::ResizePerfTestFluid/ResizePerfTest::(compare_f, 32FC1, 1, 1920x1080, 64x64, { gapi.kernel_package })                           0.036  0.021     1.76
```

The notable change can be seen only for the last case for 1080p to 64x64 resize. Which (in F32) may be a rare case. Given this, and given that there were no updates for Anna for the last year, and given that there's a lot of merge conflicts now between this branch and 4.x, I propose to close this PR for now.

**Pull request closed without merging**